cpp/src/rolling/rolling_collect_list.cuh

```diff
@@ -0,0 +1,358 @@
+  auto reduced_by_key =
+    thrust::reduce_by_key(rmm::exec_policy(stream),
+                          offsets.template begin<size_type>() + 1,  // Skip first 0 in offsets.
+                          offsets.template end<size_type>(),
+                          thrust::make_constant_iterator<size_type>(1),
+                          scatter_keys->mutable_view().template begin<size_type>(),
+                          scatter_values->mutable_view().template begin<size_type>());
+  auto scatter_values_end = reduced_by_key.second;
+  auto scatter_output =
+    make_fixed_width_column(size_data_type, num_child_rows, mask_state::UNALLOCATED, stream, mr);
+  thrust::fill_n(rmm::exec_policy(stream),
+                 scatter_output->mutable_view().template begin<size_type>(),
+                 num_child_rows,
+                 0);  // [0,0,0,...0]
+  thrust::scatter(rmm::exec_policy(stream),
+                  scatter_values->mutable_view().template begin<size_type>(),
+                  scatter_values_end,
+                  scatter_keys->view().template begin<size_type>(),
+                  scatter_output->mutable_view().template begin<size_type>());  // [0,0,1,0,0,1,...]
+
+  // Next, generate mapping with inclusive_scan() on scatter() result.
+  // For the example above:
+  //   scatter result == [0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0]
+  //   inclusive_scan == [0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4]
+  //
+  // For the case with an empty list at index 3:
+  //   scatter result == [0, 0, 1, 0, 0, 2, 0, 0, 1, 0, 0, 1, 0]
+  //   inclusive_scan == [0, 0, 1, 1, 1, 3, 3, 3, 4, 4, 4, 5, 5]
+  auto per_row_mapping =
+    make_fixed_width_column(size_data_type, num_child_rows, mask_state::UNALLOCATED, stream, mr);
+  thrust::inclusive_scan(rmm::exec_policy(stream),
+                         scatter_output->view().template begin<size_type>(),
+                         scatter_output->view().template end<size_type>(),
+                         per_row_mapping->mutable_view().template begin<size_type>());
+  return per_row_mapping;
+}
+
+/**
+ * @brief Create gather map to generate the child column of the result of
+ * the `COLLECT_LIST` window aggregation.
+ */
+template <typename PrecedingIter>
+std::unique_ptr<column> create_collect_gather_map(column_view const& child_offsets,
+                                                  column_view const& per_row_mapping,
+                                                  PrecedingIter preceding_iter,
+                                                  rmm::cuda_stream_view stream,
+                                                  rmm::mr::device_memory_resource* mr)
+{
+  auto gather_map = make_fixed_width_column(data_type{type_to_id<size_type>()},
+                                            per_row_mapping.size(),
+                                            mask_state::UNALLOCATED,
+                                            stream,
+                                            mr);
```

Comment:
I have raised #8219 to address the use of `mr` incorrectly here. These are all temporary allocations, and shouldn't be using the `mr` parameter.

I'll address this once #8158 is merged.